In [2]:
import pandas as pd
import numpy as np
import os

Para responder os exercícios a seguir, utilize os dados fornecidos na planilha "Bilateral_Country_Shares_govt_nationality_full.csv"

In [20]:
url = 'https://raw.githubusercontent.com/matheuslimaandrade/SIN-MachineLearning_Datasets/main/Bilateral_Country_Shares_govt_nationality_full.csv'
shares_gov = pd.read_csv(url, sep = ",")
shares_gov.head()

year investor_name investor  issuer_nationality_name issuer_nationality  \
0  2017     Australia      AUS                Argentina                ARG   
1  2017     Australia      AUS                Australia                AUS   
2  2017     Australia      AUS                   Brazil                BRA   
3  2017     Australia      AUS                   Canada                CAN   
4  2017     Australia      AUS  European Monetary Union                EMU   

   portfolio_share  
0         0.009804  
1         0.278169  
2         0.011028  
3         0.042183  
4         0.140563

1) Retorne em ordem crescente, os países que contem maior "portfolio_share" do Brasil nos anos de 2006, 2009 e 2013. Apresente um dataframe para cada ano mencionado (ou seja, 3 dataframes - 06', 09', 13'), contendo duas colunas: "Nome do País" e "Porfolio Share"

In [22]:
#escreva aqui
df_2006 = shares_gov[shares_gov['year'] == 2006]
df_2009 = shares_gov[shares_gov['year'] == 2009]
df_2013 = shares_gov[shares_gov['year'] == 2013]

df_2006_sorted = df_2006.sort_values(by='portfolio_share', ascending=True)
df_2009_sorted = df_2009.sort_values(by='portfolio_share', ascending=True)
df_2013_sorted = df_2013.sort_values(by='portfolio_share', ascending=True)

df_2006_result = df_2006_sorted[['issuer_nationality_name', 'portfolio_share']]
df_2009_result = df_2009_sorted[['issuer_nationality_name', 'portfolio_share']]
df_2013_result = df_2013_sorted[['issuer_nationality_name', 'portfolio_share']]

print("DataFrame para o ano de 2006:")
print(df_2006_result.head())
print()

print("DataFrame para o ano de 2009:")
print(df_2009_result.head())
print()

print("DataFrame para o ano de 2013:")
print(df_2013_result.head())



DataFrame para o ano de 2006:
     issuer_nationality_name  portfolio_share
2242            South Africa         0.000511
2103                  Mexico         0.000515
2229                  Israel         0.000551
2095               Australia         0.000555
2241               Singapore         0.000562

DataFrame para o ano de 2009:
     issuer_nationality_name  portfolio_share
1679              Luxembourg         0.000052
1665                  Cyprus         0.000093
1777                    Peru         0.000481
1622                  Greece         0.000504
1629                   Spain         0.000505

DataFrame para o ano de 2013:
    issuer_nationality_name  portfolio_share
883                   Malta         0.000002
869                  Cyprus         0.000058
881              Luxembourg         0.000213
919                 Belgium         0.000864
925             Netherlands         0.000986


2) É usual que os países se financiem por emissão de dívida interna, ou seja, que o país contenha um "porfolio_share" considerável dos seus próprios títulos relativo aos títulos de outros países. Exemplo, em 2017, ~ 27.8% da dívida nacional australiana era composta de títulos emitidos pelo próprio governo australiano. Indique no ano de 2014, qual dos países analisados possuía maior dívida interna?

In [41]:
#escreva aqui
df_2014 = shares_gov[shares_gov['year'] == 2014]

df_2014_sorted = df_2014.sort_values(by='portfolio_share', ascending=False)

maior_divida_interna = df_2014_sorted.iloc[0]

print(maior_divida_interna)

year                           2014
investor_name                Sweden
investor                        SWE
issuer_nationality_name      Sweden
issuer_nationality              SWE
portfolio_share            0.925612
Name: 718, dtype: object


Para responder os exercícios a seguir, utilize os dados fornecidos na planilha "236_data.xlsx" (na aba "Master")

In [27]:
currencies = pd.read_excel("236_data.xlsx", sheet_name= "Master", skiprows=7)

currencies.head()

Country Afghanistan Albania         Algeria Andorra Angola Anguilla  \
0  IMF Code         512     914             612    n.a.    614      312   
1       NaN         NaN     NaN             NaN     NaN    NaN      NaN   
2       NaN         NaN     NaN             NaN     NaN    NaN      NaN   
3      1946         INR     RUB             FRF     FRF    PTE      GBP   
4      1947         INR     RUB  Freely_falling     FRF    PTE      GBP   

  Antigua and Barbuda Argentina Armenia  ... Vietnam Yemen Zambia Zimbabwe  \
0                 311       213     911  ...     582   474    754      698   
1                 NaN       NaN     NaN  ...     NaN   NaN    NaN      NaN   
2                 NaN       NaN     NaN  ...     NaN   NaN    NaN      NaN   
3                 GBP       USD     RUB  ...    n.a.  n.a.    GBP      GBP   
4                 GBP       USD     RUB  ...    n.a.  n.a.    GBP      GBP   

  Czechoslovakia  USSR Yugoslavia Unnamed: 199 Unnamed: 200 Unnamed: 201  
0           n.a.  n.a.        188          NaN          NaN          NaN  
1            NaN   NaN        NaN          NaN          NaN          NaN  
2            NaN   NaN        NaN          NaN          NaN          NaN  
3            RUB   RUB        RUB          NaN          NaN          NaN  
4            RUB   RUB        RUB          NaN          NaN          NaN  

[5 rows x 202 columns]

3) Qual o ano com maior frequência de "Freely_falling" como moeda de referência?

In [38]:
#escreva aqui

years = currencies.columns[1:]
counts = currencies.set_index('Country')[years].apply(pd.Series.value_counts, axis=1)['Freely_falling'].fillna(0)
sorted_counts = counts.sort_values(ascending=False)
most_common_year = sorted_counts.idxmax()
print(most_common_year)




1993


4) Com a queda da União Soviética, além da separação política, foi possível observar uma movimentação na moeda utilizada como âncora. Retorne um dataframe que contenha o ano em que cada país da lista abaixo deixou de utilizar o RUB (ruble)

In [ ]:
#Para ajudar, aqui está uma lista com os nomes dos países (ex-Russia) que deve ser utilizada p este exercício
urss_list = ["Ukraine", "Georgia", "Uzbekistan", "Armenia", "Azerbaijan", "Kazakhstan", "Kyrgyzstan", "Moldova",
"Turkmenistan", "Tajikistan", "Latvia", "Lithuania", "Estonia"]

In [50]:
selected_df = currencies[currencies.columns.intersection(urss_list)]
selected_df_years = pd.concat([year_col, selected_df], axis=1).set_index("Country")

years = []
for country in urss_list:
    selected_df_years[country] = selected_df_years[country].ne(selected_df_years[country].shift().bfill()).astype(int)
    year_y = selected_df_years[country].idxmax()
    years.append(year_y)

final_df = pd.DataFrame({"Country":urss_list, "Years" : years})

print(final_df)

         Country  Years
0        Ukraine    NaN
1        Georgia    NaN
2     Uzbekistan    NaN
3        Armenia    NaN
4     Azerbaijan    NaN
5     Kazakhstan    NaN
6     Kyrgyzstan    NaN
7        Moldova    NaN
8   Turkmenistan    NaN
9     Tajikistan    NaN
10        Latvia    NaN
11     Lithuania    NaN
12       Estonia    NaN


5) Retorne a década em que houve maior troca de "Currency Anchors" no mundo

In [54]:
#escreva aqui
currencies_indexed = currencies.set_index('Country')
countries_list = currencies.columns[1:199]
for country in countries_list:
    # Determine if there is a change in currency from the previous year
    currencies_indexed[country] = currencies_indexed[country].ne(currencies_indexed[country].shift().bfill()).astype(int)
currencies_indexed['change'] = currencies_indexed.sum(axis=1)
year_x = currencies_indexed['change'].idxmax()
year_x


<ipython-input-54-cb35877a4e33>:7: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  currencies_indexed['change'] = currencies_indexed.sum(axis=1)


1999